File: port/swift/Sources/MuonLedger/PrintCommand.swift
/// Print command: re-output parsed transactions in standard ledger format.
///
/// Outputs transactions in a clean, canonical format matching C++ ledger's
/// print command output.

import Foundation

// MARK: - PrintCommand

/// Options for the print command.
struct PrintOptions {
    var patterns: [String] = []
}

/// Format a date as YYYY/MM/DD for print output.
private func formatPrintDate(_ date: Date?) -> String {
    guard let date = date else { return "" }
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = TimeZone(identifier: "UTC")!
    let year = calendar.component(.year, from: date)
    let month = calendar.component(.month, from: date)
    let day = calendar.component(.day, from: date)
    return String(format: "%04d/%02d/%02d", year, month, day)
}

/// Generate a print report from a journal.
func printCommand(journal: Journal, opts: PrintOptions) -> String {
    var output = ""

    for xact in journal.transactions {
        // Date line
        var dateLine = formatPrintDate(xact.date)

        // State marker
        switch xact.state {
        case .cleared:
            dateLine += " *"
        case .pending:
            dateLine += " !"
        case .uncleared:
            break
        }

        // Payee
        dateLine += " \(xact.payee)"

        output += dateLine + "\n"

        // Postings
        for post in xact.posts {
            guard let acct = post.account else { continue }
            let accountName = acct.fullname

            let amtStr: String
            if let amt = post.amount, !amt.isNull {
                amtStr = amt.toString()
            } else {
                amtStr = ""
            }

            if amtStr.isEmpty {
                output += "    \(accountName)\n"
            } else {
                // Right-align amount at roughly column 48
                let acctWidth = 40
                if accountName.count < acctWidth {
                    let padding = String(repeating: " ", count: acctWidth - accountName.count)
                    output += "    \(accountName)\(padding)  \(amtStr)\n"
                } else {
                    output += "    \(accountName)  \(amtStr)\n"
                }
            }
        }

        output += "\n"
    }

    return output
}

File: port/kotlin/src/main/kotlin/muonledger/PrintCommand.kt
package muonledger

/**
 * Print command: re-output transactions in ledger format.
 */
object PrintCommand {

    fun run(journal: Journal): String {
        val sb = StringBuilder()

        for (xact in journal.transactions) {
            // Date
            val date = xact.date
            if (date != null) {
                sb.append("%04d/%02d/%02d".format(date.year, date.monthValue, date.dayOfMonth))
            }

            // Payee
            sb.append(" ${xact.payee}\n")

            // Postings
            for (post in xact.posts) {
                val acctName = post.account.fullname
                val amtStr = post.amount?.let {
                    if (!it.isNull) it.toString() else null
                }

                if (amtStr != null) {
                    // Right-align amount at column 48 (4 indent + 40 account + space)
                    val line = "    ${acctName.padEnd(40)}  $amtStr"
                    sb.append(line).append('\n')
                } else {
                    sb.append("    $acctName\n")
                }
            }

            sb.append('\n')
        }

        return sb.toString()
    }
}
